theano/sparse/opt.py

```diff
@@ -1892,15 +1896,27 @@ def c_code(self, node, name, inputs, outputs, sub):
             memcpy(Dzi, Dpi, PyArray_DIMS(%(p_ind)s)[0]*sizeof(dtype_%(p_ind)s));
             memcpy(Dzp, Dpp, PyArray_DIMS(%(p_ptr)s)[0]*sizeof(dtype_%(p_ptr)s));
 
+            // blas expects ints; convert here (rather than just making K etc ints) to avoid potential overflow in the negative-stride correction
+            int K32 = K;
+            int Sdx32 = Sdx;
+            int Sdy32 = Sdy;
+
             for (npy_int32 m = 0; m < M; ++m) {
                 for (npy_int32 n_idx = Dpp[m * Sdpp]; n_idx < Dpp[(m+1)*Sdpp]; ++n_idx) {
                     const npy_int32 n = Dpi[n_idx * Sdpi]; // row index of non-null value for column K
 
                     const dtype_%(x)s* x_row = (dtype_%(x)s*)(PyArray_BYTES(%(x)s) + PyArray_STRIDES(%(x)s)[0] * m);
 
                     const dtype_%(y)s* y_col = (dtype_%(y)s*)(PyArray_BYTES(%(y)s) + PyArray_STRIDES(%(y)s)[0] * n);
+                    // dot expects pointer to the beginning of memory arrays,
+                    // so when the stride is negative, we need to get the
+                    // last element
+                    if (Sdx < 0)
+                        x_row += (K - 1) * Sdx;
+                    if (Sdy < 0)
+                        y_col += (K - 1) * Sdy;
```

Comment:
The stride stays negative because that's how BLAS knows to start from the last-in-memory element.

Debian has [already applied these fixes locally](https://anonscm.debian.org/cgit/debian-science/packages/theano.git/commit/?id=7f0a0528eb4802eb3e0cbec47bbd1198a3a44403): they have a "packages must build on as many of our supported architectures (which include big-endian ones) as reasonably possible" policy, which makes the second part a worse bug for them than for x86-only users.

I'd consider the first part the worse bug, since it produces silently wrong answers, but don't know how likely a negative-stride array (followed by memory the Theano process also owns, such as another array - if it doesn't, you'll get a segfault, which at least lets you know *something* is wrong) is in practice.